docs/quantum_gate.rst

```diff
@@ -1,4 +1,366 @@
++---------------+           +-----------------+            +            +----------------+
+| angle         |           | numpy.pi        |            |            | double         |
++---------------+           +-----------------+            +            +----------------+
+| type          |           | __t_gate__      |            |            | gate_type_t    |
++---------------+-----------+-----------------+------------+            +----------------+
+| duration      | semantic  | 80              | schedulers,|            | size_t         |
+|               |           |                 | etc.       |            |                |
++---------------+           +-----------------+------------+            +----------------+
+| mat           |           |                 | optimizer  |            | cmat_t         |
+|               |           |                 | pass       |            |                |
++---------------+-----------+-----------------+------------+------------+----------------+
+| cycle         | result    | 4               | code       | scheduler  | size_t         |
+|               |           |                 | generation |            |                |
++---------------+-----------+-----------------+------------+------------+----------------+
+
+Custom gates have an additional set of attributes,
+primarily supporting the initialization of the gate attributes from configuration file parameters.
+
+Some further notes on the gate attributes:
+
+- the name of a gate includes the string representations of its qubit operands in case of a specialized gate;
+  so in general, when given a name, one has to take care to isolate the operation from it;
+  one may assume that the operation is a single identifier optionally followed by white space and the operands
+
+- gates are most directly distinguished by their name
+
+:Note: Distinguishing gates internally in the compiler by their name is problematic; distinguishing by their type (see the table below) would be better; the latter conveys the semantic definition and is independent of the representation (e.g. ``mrx90``, ``mx90``, and ``Rmx90`` all could be names of a -90 degrees X rotation); furthermore, a name is something of the external representation and is mapped to the internal representation using the platform configuration file; however, the enumeration type of type can never include all possible gates (e.g. those with arbitrary angles, any number of operands, etc.) so the type inevitably can be imprecise; but it can be precise when the type refers to the operation only, i.e. excluding the operands
+
+- qubit and classical operands are represented by unsigned valued indices starting from 0 in their respective register spaces
+
+- ``angle`` is in radians; it specifies the value of the arbitrary angle of those operations that need one; it is initialized only from an explicit specification as parameter value of a gate creation API
```

Comment:
I understand why this is done, but this makes it a bit tricky to specify common angles like 180 or 90 degrees (given that there is no definition of Pi available) .

Reply:
Yes, this has been a choice.
We could as well have defined an attribute pi_fraction instead,
so that the angle would have become pi_fraction*pi
where pi would have been built-in into OpenQL.